### DISCUSSION TITLE
Creating the command inputs for ChimeraX

### DISCUSSION THREAD

**Discussion by roshan-eng** [jitinsingla/TetraScape#4] [General]
Currently the inputs are taken form ChimeraX shell in a list and dictionary format. I don't have the idea of variable types ChimeraX supports to integrate this input via. ChimeraX command line. The current input requirements is in line 384.

@tomgoddard , please look if you can fix this.

**Comment by roshan-eng** [comment]
1. While looking at atoms in residues.atoms for ASN, i am getting ['N', 'CA', 'C', 'O']. How to get the c_beta chain in it ? 
2. Sometime i get residues named like "9DK" with atom constituents like ['C67', 'C69', 'C70', 'C71', 'C72', 'C75', 'C76', 'C77', 'C78', 'C80', 'N74', 'N79', 'N81', 'O68', 'O73']. What are these and how can make a check to skip these not-amino acid components reliably ?

@tomgoddard

**Reply by tomgoddard** [reply]
The coordinates for any atoms of a residue may be missing.  Often the side-chain atoms like C-beta are missing because the experimental data was of poor quality so they could not place the side chain.  Some very poor quality data will have structures that only have the C-alpha atom for each residue.  Also GLY has no C-beta atom.  So hopefully you can still place tetrahedrons when the C-beta atom position is not available.

To make sure you only place tetrahedra for amino acids you should check the residue type to make sure it is an amino acid.  I sent you code to do that 5 months ago but I don't see it in the current GitHub code.  Here is the email and I will attach the code in a zip file.

> From: Tom Goddard
> Subject: Re: Command for Massing
> Date: July 18, 2023 at 1:34:27 PM PDT
> To: Jitin Singla 
> Cc: Helen Berman
> 
> Hi Jitin,
> 
>   I fixed up the ChimeraX tetra and massing commands so they work with chain arguments (e.g. "tetra /A,B" or "massing /C") making a new file Tetra2.py attached to this email.  For the chains argument I used UniqueChainsArg.  Also I changed the code that checks if a residue is an amino acid and changed the code that finds the N, C, CA, CB atoms in the residue so that those work more reliably.  I've also attached Tetra2.diffs so you can more easily see the lines I changed.
> 
>   Here are some further steps I suggest your student make.
> 
>   1)  The Tetra class looks through all the currently open ChimeraX models.  I'd suggest instead that Tetra not look at the open models and only operate on the chains handed to the tetrahedron() and massing() functions.  If you want the commands to operate on all chains when none are specified then find all the chains in the tetrahedron_model() and massing_model() functions.
> 
>   2) The massing() function takes a list of chain identifiers (e.g. A, B, ..).  It should instead take a list of ChimeraX chain objects.  Multiple models can be opened that have the same chain identifiers so just knowing the chain identifier (e.g. A) is not enough.
> 
>   3) It will be useful if the tetrahedron() and massing() functions can work on just a subset of residues.  So probably those routines should take a list of residues instead of chains.  From the residues the code can figure out the chains and might want to compute the depiction for the entire chain but just limit the output to the specified residues.
> 
>   If you give the code to the USC people be sure to tell them they need to install alphashape in ChimeraX with ChimeraX command "pip install alphashape" to use the massing command.  Maybe also they will need to install trimesh ("pip install trimesh") -- I see the code sometimes uses that, although it did not need it for my test cases.
> 
>   Please ask more questions and always provide the latest code.
> 
> 	Tom

[tetra.zip](https://github.com/jitinsingla/TetrahedralRepresentation/files/13747362/tetra.zip)

**Reply by roshan-eng** [reply]
Thank you. Well you might be referencing to #L165 in Tetra.py.
There is one more problem i am struggling with. I want to give  a sequence of amino based on its numbering that which are needed to be in massing and which are to be in tetrahedrons model. The problem is that i need to know the last amino number (residue.number) that will define my sequence. HELP ! @tomgoddard

**Reply by tomgoddard** [reply]
I see there is a check for residue type being amino acid in Tetra.py.  That should exclude the residue "9DK" you mentioned since that is a small molecule ligand (not an amino acid): https://www.rcsb.org/ligand/9DK

**Comment by tomgoddard** [comment]
In order to make suggestions about how to improve the code it needs to work at least on a simple example.  Otherwise I cannot test things and am left guessing what you need.

**Reply by roshan-eng** [reply]
Hey @tomgoddard , I completed the code formatting. Its working fine from my side. How to use it for now is in README. I also added some comments. If anything comes up then you can comment here tagging me along.

**Reply by roshan-eng** [reply]
Here's the code running on 5VAI.
> from TetraModelling.main import Tetra
> tetra = Tetra(session)
> tetra.massing(sequence={"A":[(0, 100)], "B":[(20, 40), (60, 100)], "R":[(240, 300), (350, 420)]}, unit=1.5)
![image3](https://github.com/jitinsingla/TetrahedralRepresentation/assets/77035385/14ea1173-16f4-4291-a255-c4b5ade68d17)

**Comment by roshan-eng** [comment]
Hey @tomgoddard , is the code working as expected from your side ? I might be busy on some other stuff for few days so before that i want to make it working for you to continue on it.

**Reply by tomgoddard** [reply]
The tetrascape_cmd code worked fine for me just now.  It did not color the tetrahedra.  Maybe that is only in the Tetra.py copy of the code.  It is important that there is only one copy of the code, not two copies that are out of sync with each other.

**Comment by tomgoddard** [comment]
Hi Roshan,  I will try it a few days from now. If it works for you it will work for me. Then I can help with making a ChimeraX command  I am on Christmas and New Year's holiday now, so not working. Tom

**Reply by roshan-eng** [reply]
Great! Merry christmas and Happy New year to you : )

**Comment by tomgoddard** [comment]
Ok, I fixed the massing ChimeraX command code in Tetra.py.  Here it is.

[tetra.zip](https://github.com/jitinsingla/TetrahedralRepresentation/files/13847618/tetra.zip)

To compute massing for your 5VAI example you would use ChimeraX commands

cd /directory/containing/Tetra.py
open Tetra.py
open 5vai
mass /A:1-100 /B:20-40,60-100 /R:240-300,350-420 unit 1.5

I have several further suggestions.  I will make new github issues for those.

**Comment by roshan-eng** [comment]
@jitinsingla, I added the functionality to filter the residues based on secondary structures for tetrahedron function.

**Comment by roshan-eng** [comment]
Hey @tomgoddard , I had one parameter of "ss" in,
https://github.com/jitinsingla/TetrahedralRepresentation/blob/32b692774f1522000dd6bbaff8127578da1feba0/tetrascape_cmd/src/cmd.py#L296

How can i integrate this with command line input. Like is there's any list input method ?

**Comment by tomgoddard** [comment]
Here is how you can handle the ss (secondary structure) argument in the ChimeraX command

tetra /A ss helix,strand

```
	@staticmethod
	def register_command(bi, ci, logger):

            from chimerax.core.commands import BoolArg, IntArg, CmdDesc, register, FloatArg, StringArg
            from chimerax.atomic import ResiduesArg
	    t_desc = CmdDesc(required = [], 
	    				optional = [("residues", ResiduesArg), ("revSeq", BoolArg)], 
	    				keyword = [("ss", StringArg)],
	    				synopsis = 'creates tetrahedral model')
	    register('tetra', t_desc, cmd.tetrahedral_model, logger=logger)
```


```
def tetrahedral_model(session, residues=None, revSeq=False, ss=['all']):
    if residues is None:
        from chimerax.atomic import all_residues
        residues = all_residues(session)
    if isinstance(ss, str):
        ss = ss.split(',')
    for structure, chain_residue_intervals in residue_intervals(residues):
        t = Tetra(session, models = [structure])
        t.tetrahedron(sequence = chain_residue_intervals, in_sequence = not revSeq)

```
